File: Problems/Algorithms/94. Binary Tree Inorder Traversal/InorderTraversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        val results = ArrayList<Int>()
        
        if (root != null) {
           results.addAll(inorderTraversal(root.left))
           results.add(root.`val`)
           results.addAll(inorderTraversal(root.right))
        }
        
        return results
    }
}

File: Problems/Algorithms/94. Binary Tree Inorder Traversal/inorder_traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut res = vec![];
        Self::helper(root, &mut res);
        res
    }

    fn helper(root: Option<Rc<RefCell<TreeNode>>>, res: &mut Vec<i32>) {
        match root {
            None => (),
            Some(node) => {
                let node = node.borrow();
                Self::helper(node.left.clone(), res);
                res.push(node.val);
                Self::helper(node.right.clone(), res);
            }
        }
    }
}
